File: packages/vm/wasmlib/ts/wasmclient/decoder.ts
// The Results struct is used to gather all arguments for a smart
// contract function call and encode it into a deterministic byte array
import * as wasmclient from "./index";
import {Base58} from "./crypto";
import {Buffer} from "./buffer";

export class Decoder {
    private static checkDefault(bytes: Buffer | undefined, typeID: wasmclient.Int32): Buffer {
        const size = wasmclient.TYPE_SIZES[typeID];
        if (bytes === undefined) {
            return Buffer.alloc(size);
        }
        if (size != 0 && bytes.length != size) {
            wasmclient.panic("invalid type size");
        }
        return bytes;
    }

    private static toBase58(bytes: Buffer | undefined, typeID: wasmclient.Int32): string {
        return Base58.encode(Decoder.checkDefault(bytes, typeID));
    }

    protected toAddress(bytes: Buffer | undefined): wasmclient.Address {
        return Decoder.toBase58(bytes, wasmclient.TYPE_ADDRESS);
    }

    protected toAgentID(bytes: Buffer | undefined): wasmclient.AgentID {
        return Decoder.toBase58(bytes, wasmclient.TYPE_AGENT_ID);
    }

    protected toBytes(bytes: Buffer | undefined): wasmclient.Bytes {
        return Decoder.checkDefault(bytes, wasmclient.TYPE_BYTES)
    }

    protected toBool(bytes: Buffer | undefined): wasmclient.Bool {
        return Decoder.checkDefault(bytes, wasmclient.TYPE_BOOL)[0] != 0;
    }

    protected toChainID(bytes: Buffer | undefined): wasmclient.ChainID {
        return Decoder.toBase58(bytes, wasmclient.TYPE_CHAIN_ID);
    }

    protected toColor(bytes: Buffer | undefined): wasmclient.Color {
        let color = Decoder.toBase58(bytes, wasmclient.TYPE_COLOR);
        if (color == "11111111111111111111111111111111") {
            color = "IOTA";
        }
        return color;
    }

    protected toHash(bytes: Buffer | undefined): wasmclient.Hash {
        return Decoder.toBase58(bytes, wasmclient.TYPE_HASH);
    }

    protected toHname(bytes: Buffer | undefined): wasmclient.Hname {
        return Decoder.checkDefault(bytes, wasmclient.TYPE_HNAME).readUInt32LE(0);
    }

    protected toInt8(bytes: Buffer | undefined): wasmclient.Int8 {
        return Decoder.checkDefault(bytes, wasmclient.TYPE_INT8).readInt8(0);
    }

    protected toInt16(bytes: Buffer | undefined): wasmclient.Int16 {
        return Decoder.checkDefault(bytes, wasmclient.TYPE_INT16).readInt16LE(0);
    }

    protected toInt32(bytes: Buffer | undefined): wasmclient.Int32 {
        return Decoder.checkDefault(bytes, wasmclient.TYPE_INT32).readInt32LE(0);
    }

    protected toInt64(bytes: Buffer | undefined): wasmclient.Int64 {
        return Decoder.checkDefault(bytes, wasmclient.TYPE_INT64).readBigInt64LE(0);
    }

    protected toRequestID(bytes: Buffer | undefined): wasmclient.RequestID {
        return Decoder.toBase58(bytes, wasmclient.TYPE_REQUEST_ID);
    }

    protected toString(bytes: Buffer | undefined): wasmclient.String {
        return Decoder.checkDefault(bytes, wasmclient.TYPE_STRING).toString();
    }

    protected toUint8(bytes: Buffer | undefined): wasmclient.Uint8 {
        return Decoder.checkDefault(bytes, wasmclient.TYPE_INT8).readUInt8(0);
    }

    protected toUint16(bytes: Buffer | undefined): wasmclient.Uint16 {
        return Decoder.checkDefault(bytes, wasmclient.TYPE_INT16).readUInt16LE(0);
    }

    protected toUint32(bytes: Buffer | undefined): wasmclient.Uint32 {
        return Decoder.checkDefault(bytes, wasmclient.TYPE_INT32).readUInt32LE(0);
    }

    protected toUint64(bytes: Buffer | undefined): wasmclient.Uint64 {
        return Decoder.checkDefault(bytes, wasmclient.TYPE_INT64).readBigUInt64LE(0);
    }
}

File: packages/wasmvm/wasmlib/go/wasmclient/decoder.go
package wasmclient

import "encoding/binary"

type Decoder struct{}

func checkDefault(bytes []byte, typeID int32) []byte {
	size := TypeSizes[typeID]
	if bytes == nil {
		// return default all-zero bytes value
		return make([]byte, size)
	}
	if size != 0 && len(bytes) != int(size) {
		panic("invalid type size")
	}
	return bytes
}

func toBase58(bytes []byte, typeID int32) string {
	return Base58Encode(checkDefault(bytes, typeID))
}

func (c Decoder) ToAddress(bytes []byte) Address {
	return Address(toBase58(bytes, TYPE_ADDRESS))
}

func (c Decoder) ToAgentID(bytes []byte) AgentID {
	return AgentID(toBase58(bytes, TYPE_AGENT_ID))
}

func (c Decoder) ToBool(bytes []byte) bool {
	return checkDefault(bytes, TYPE_BOOL)[0] != 0
}

func (c Decoder) ToBytes(bytes []byte) []byte {
	return checkDefault(bytes, TYPE_BYTES)
}

func (c Decoder) ToChainID(bytes []byte) ChainID {
	return ChainID(toBase58(bytes, TYPE_CHAIN_ID))
}

func (c Decoder) ToColor(bytes []byte) Color {
	color := toBase58(bytes, TYPE_COLOR)
	if color == COLOR_IOTA_BASE58 {
		color = COLOR_IOTA
	}
	return Color(color)
}

func (c Decoder) ToHash(bytes []byte) Hash {
	return Hash(toBase58(bytes, TYPE_HASH))
}

func (c Decoder) ToHname(bytes []byte) Hname {
	return Hname(binary.LittleEndian.Uint32(checkDefault(bytes, TYPE_HNAME)))
}

func (c Decoder) ToInt8(bytes []byte) int8 {
	return int8(c.ToUint8(bytes))
}

func (c Decoder) ToInt16(bytes []byte) int16 {
	return int16(c.ToUint16(bytes))
}

func (c Decoder) ToInt32(bytes []byte) int32 {
	return int32(c.ToUint32(bytes))
}

func (c Decoder) ToInt64(bytes []byte) int64 {
	return int64(c.ToUint64(bytes))
}

func (c Decoder) ToRequestID(bytes []byte) RequestID {
	return RequestID(toBase58(bytes, TYPE_REQUEST_ID))
}

func (c Decoder) ToString(bytes []byte) string {
	return string(checkDefault(bytes, TYPE_STRING))
}

func (c Decoder) ToUint8(bytes []byte) uint8 {
	return checkDefault(bytes, TYPE_INT8)[0]
}

func (c Decoder) ToUint16(bytes []byte) uint16 {
	return binary.LittleEndian.Uint16(checkDefault(bytes, TYPE_INT16))
}

func (c Decoder) ToUint32(bytes []byte) uint32 {
	return binary.LittleEndian.Uint32(checkDefault(bytes, TYPE_INT32))
}

func (c Decoder) ToUint64(bytes []byte) uint64 {
	return binary.LittleEndian.Uint64(checkDefault(bytes, TYPE_INT64))
}
